lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -88,38 +191,35 @@ int nrf_modem_lib_trace_start(enum nrf_modem_lib_trace_mode trace_mode)
 
 int nrf_modem_lib_trace_process(const uint8_t *data, uint32_t len)
 {
-	if (!is_transport_initialized) {
-		return -ENXIO;
-	}
+	int ret = 0;
 
-#ifdef CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART
-	/* Split RAM buffer into smaller chunks to be transferred using DMA. */
-	uint32_t remaining_bytes = len;
-	const uint32_t MAX_BUF_LEN = (1 << UARTE1_EASYDMA_MAXCNT_SIZE) - 1;
+	if (is_transport_initialized) {
+		struct trace_data_t trace_data = { .data = data, .len = len };
+		char *mem_ptr = k_malloc(sizeof(trace_data));
 
-	while (remaining_bytes) {
-		size_t transfer_len = MIN(remaining_bytes, MAX_BUF_LEN);
-		uint32_t idx = len - remaining_bytes;
+		if (mem_ptr != NULL) {
+			memcpy(mem_ptr, &trace_data, sizeof(trace_data));
 
-		nrfx_uarte_tx(&uarte_inst, &data[idx], transfer_len);
-		remaining_bytes -= transfer_len;
-	}
-#endif
+			k_fifo_put(&trace_fifo, mem_ptr);
+		} else {
+			LOG_ERR("k_malloc failed.");
 
-#ifdef CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT
-	uint32_t remaining_bytes = len;
+			ret = -ENOMEM;
+		}
+	} else {
+		LOG_ERR("Modem trace received without initialzing transport");
 
-	while (remaining_bytes) {
-		uint8_t transfer_len = MIN(remaining_bytes,
-					CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT_BUF_SIZE);
-		uint32_t idx = len - remaining_bytes;
+		ret = -ENXIO;
+	}
+
+	if (ret) {
+		int err = nrf_modem_trace_processed_callback(data, len);
```

Comment:
This should be done unconditionally, without checking `ret` (which can't be zero, right?)

Reply:
`ret` will be 0 if the transport is initialized and there is enough heap space. Then the thread will be responsible for calling the `nrf_modem_trace_processed_callback`. So here it's only calling the callback in case of an error.

Reply:
Ok, thanks for clarifying. I suggest using a `goto` pattern to call `nrf_modem_trace_processed_callback`; it's easier to follow than manipulating one variable and acting on it at the end of a function.
```
if (!initialized) {
    goto trace_processed;
}

fifo_item = k_malloc(...);
if (!fifo_item) {
    goto trace_processed;
}

fifo_put();
return;

trace_processed:
    ....
}
```

Reply:
I prefer not to use goto statement (I believe they have been frowned upon since 1960s and is also one of the MISRA-2012 rules that unfortunately Zephyr does not enforce yet due to legacy code containing of bunch of those goto statements. See https://rules.sonarsource.com/c/RSPEC-907 and https://stackoverflow.com/a/46789/5084499 among many other resources ). It just makes the code hard to read.  

In our case, I believe what is confusing is the fact that we use `ret` to check if something failed in an earlier part of this function. Perhaps renaming `ret` to `success` would be better?

Reply:
I agree with @balaji-nordic that `goto` should be avoided. Think the main was the mixing of `ret` and `err`. Refactored it a bit now to improve readability. Does it look ok @lemrey ?